client_code/pedantic.py

```diff
@@ -0,0 +1,56 @@
+# SPDX-License-Identifier: MIT
+# Copyright (c) 2021 anvilistas
+
+__version__ = "0.0.1"
+
+
+class BaseValidator:
+    value = None
+
+    def is_valid(self):
+        raise NotImplementedError
+
+    def __get__(self, obj, objtype=None):
+        return getattr(obj, self.private_name)
+
+    def __set__(self, obj, value):
+        self.value = value
+        if not self.is_valid():
+            raise ValueError(f"{value} is not a valid value")
+        setattr(obj, self.private_name, value)
+
+
+def _is_valid(obj):
+    return all(v() for v in obj.validators)
+
+
+def validate(**kwargs):
+    def decorator(cls):
+        cls.is_valid = _is_valid
+
+        for attr, validator in kwargs.items():
+            if getattr(cls, attr, None) is not None:
+                raise ValueError(f"{attr} is already defined")
```

Comment:
One thing I was thinking was that this prevents using a pattern of multiple validators. Not sure that's something we want to support but I can envisage wanting to use a few validators for a single property.

Reply:
I suppose you could write a bespoke validator that 'chained' some of the standard validators  fairly easily.

Reply:
maybe I didn't post it - I was thinking `combine` might be a nice helper function (or `chain`)
`@validate(foo=combine(InList(x), IsInstance(str)))`